### DISCUSSION TITLE
What are “templates” and should we remove or fix them?

### DISCUSSION THREAD

**Discussion by cassidyjames** [elementary/code#947] [General]
I'm finding references to a [TemplateManager](https://github.com/elementary/code/blob/ceb3b1f34f6ab4d8ea242c0c46e5a126d48846f8/src/Services/TemplateManager.vala) and templates throughout Code, but I've literally never seen them and have no idea what it's supposed to be. So my question for discussion is:

1. What are they supposed to be?
2. Should we remove them, or fix them?

---

https://github.com/elementary/code/blob/ceb3b1f34f6ab4d8ea242c0c46e5a126d48846f8/src/Services/TemplateManager.vala#L1-L268

https://github.com/elementary/code/blob/ceb3b1f34f6ab4d8ea242c0c46e5a126d48846f8/src/Services/PluginManager.vala#L35-L41

https://github.com/elementary/code/blob/7b6e3313aefef1c55f1d8cf65137bee86d376dee/src/MainWindow.vala#L255-L258

https://github.com/elementary/code/blob/7b6e3313aefef1c55f1d8cf65137bee86d376dee/src/MainWindow.vala#L905-L907

https://github.com/elementary/code/blob/aff18f8ce3c0642ab426cf6124b234db8caadfd9/src/Widgets/HeaderBar.vala#L52-L54

**Comment by pofart** [comment]
Hello @cassidyjames !
I wish there was a feature like templates, from time to time I write something like simple bug reports, and it would speed up my work a bit, since my main text editor is yours Code! 
Thanks for your works!

**Comment by jeremypw** [comment]
I would be willing to complete this function if thought useful for Code's current purpose.  As far as I can tell the intention is to have templates containing keyed placeholders that are replaced by variable text on creating a new document from them.  However there is no UI for creating templates of this type nor for using them and for specifying what text replaces the placeholders.

Much the same functionality is already available by copying a file from the existing home Templates folder and using Search/Replace to fill in placeholders.

**Reply by zeebok** [reply]
I would be happy with some kind of shortcut or menu to load in a file from the Templates folder as a new file or paste the contents into the current document to be used as a "snippet" of sorts. And I guess it would be nice to be able to save a highlighted piece of code as a snippet to the Templates folder.

**Reply by jeremypw** [reply]
There is already a menu option "Templates" under "New" in the folder context menu in the sidebar but this only appears if there are text files in ~/Templates.  This is intended for creating new files rather than inserting snippets which is a whole different issue - see #36. This will probably be addressed after Gtk4 porting.

The idea of saving a file or selection to a file in Templates is interesting - perhaps raise an issue about this.
